### ISSUE TITLE
Support _FILE based configuration values for all config keys

### ISSUE THREAD

**Issue by m90** [offen/docker-volume-backup#253]
To configure this tool, a lot of potentially sensitive and powerful credentials and tokens need to be passed. Right now, these have to be passed as env vars. Docker Swarm supports "secrets" which allows putting credentials into files made available to the container at runtime.

This was added for S3 credentials in https://github.com/offen/docker-volume-backup/pull/161, but not for anything else.

Unfortunately package `envconfig` which is used here does not (and will not) support this out of the box: https://github.com/kelseyhightower/envconfig/issues/130

Supporting this pattern for all values would be a very nice addition though, so I can see a few ways forward:
- find out if there is a fork of package `envconfig` that supports this (it's mostly dormant anyways)
- fork package `envconfig` and add the functionality ourselves
- find a different package for handling configuration (this would need to be non-breaking)

**Comment by m90** [comment]
There seems to be https://github.com/johnstairs/pathenvconfig

**Comment by MaxJa4** [comment]
> There seems to be https://github.com/johnstairs/pathenvconfig

Just had a look at that code. It doesn't allow to specify a custom name like `envconfig:"EMAIL_SMTP_PASSWORD"`, which is needed to process names with multiple capitalized letters next to each other. There also seems to be a bug in the used regex code. Wouldn't use it.

I found this though: https://github.com/caarlos0/env
Looks promising, will try it out.

**Comment by m90** [comment]
The nice thing about https://github.com/johnstairs/pathenvconfig compared to https://github.com/caarlos0/env is that you don't need to duplicate configuration keys. It just automatically picks up `_FILE` values when they're there, falling back to plain env vars when they're not. Ideally, we'd just support that for _everything_ and hiding the entire logic in the configuration library.

**Comment by MaxJa4** [comment]
That would be ideal, yes. But that particular package doesn't work with capitalized variable names like SMTP and while trying out the regex (to verify that the cap. variables do/don't work) I found a regex error, which should make it useless (unless the regex tool and I are wrong).
Anyway, maybe we can fork it ourselves and just fix those two issues?

**Comment by m90** [comment]
Question is whether it's easier to make `env` or `envconfig` do what we want. Right now, I'd lean towards envconfig as we'd only have to teach it a single trick. Then again, I still cannot believe someone else has done this yet.

**Comment by MaxJa4** [comment]
`env` is more actively maintained, but `envconfig` has worked well for a while in this project.
I already implemented `env` and it ran through the CI successfully here: https://github.com/offen/docker-volume-backup/pull/264/commits/cc4492d2c0500c50e561d31ab2b42bc8aea30c0c
Now I'm also checking out `pathenvconfig` just to be sure that it doesn't work for us.

**Comment by MaxJa4** [comment]
@m90 Would you prefer to fork it yourself or in the offen organization? Otherwise I will do that later today, as I would volunteer to add that to ’envconfig’.

**Comment by m90** [comment]
Depends on if you wanna maintain it in case someone else starts using it :) Maybe it'd make sense if I forked it to the Offen org and made you a repo admin? Let me know if that works for you. We also use the package in `offen/offen` where this feature has also been requested, so that'd be nice as well.

As for the implementation itself, I think it would make sense to keep this as low-profile as possible (as not to create maintenance burden) and use an approach similar to the one described in this comment: https://github.com/kelseyhightower/envconfig/issues/130#issuecomment-646301226

I think you wouldn't even need the new constructor and instead allow overriding a global lookup function that could then check whether `_FILE` exists:

```go
        envconfig.Lookup = func(info VarInfo) (bool, string) { /* custom logic */ }
	if err := envconfig.Process("", s.c); err != nil {
		return nil, fmt.Errorf("newScript: failed to process configuration values: %w", err)
	}
```

What do you think?

**Comment by m90** [comment]
> You mean the logic for reading the file would be inside /* custom logic */ instead of inside the package? Meaning, that only lookupEnv can be changed by .Lookup from e.g. os.LookupEnv to that custom lookup function?

Yes, I think that should work and would create the least invasive change in the original package. Also, reading the comments (https://github.com/kelseyhightower/envconfig/issues/130#issuecomment-648832012), it should be the only change that at least has minimal changes of getting merged back upstream.

> I'd opt to remove all of this unused legacy stuff.

Depends on if the goal is to get the change merged, or to keep the fork separate. Not sure right now. Another option would be  to just copy over the parts of the package we need, change them and keep them in this repo.

> Also, there is an issue of just reading anything as file and using its contents which ends with _FILE

Not sure if I understand every detail, but if you'd implement it using a custom lookup function like:

```go
func fileAwareLookup(key string) (string, bool) {
	location, ok := os.LookupEnv(key + "_FILE")
	if ok {
		contents, err := os.ReadFile(location)
		if err != nil {
			return "", false
		}
		return string(contents), true
	}
	return os.LookupEnv(key)
}
```

`SSH_IDENTITY_FILE`  would just become `SSH_IDENTITY_FILE_FILE`. That's maybe a bit confusing, but works and can be cushioned by some docs.

**Comment by MaxJa4** [comment]
> Yes, I think that should work and would create the least invasive change in the original package. Also, reading the comments (https://github.com/kelseyhightower/envconfig/issues/130#issuecomment-648832012), it should be the only change that at least has minimal changes of getting merged back upstream.

I think the statement of the repo owner in this issue is very clear:

> Custom lookup functions make sense to me and I think it's the right path forward.

So only adding the `Lookup` function would probably have good chances of being merged to the original repo which removes a "burden" on our side again (maintaining another project).

> Depends on if the goal is to get the change merged, or to keep the fork separate. Not sure right now. Another option would be to just copy over the parts of the package we need, change them and keep them in this repo.

Good point. If we want it to get merged, then definitely don't change anything else than the `Lookup` function. If we just want to integrate it in the codebase here, we'd only need [envconfig.go](https://github.com/kelseyhightower/envconfig/blob/master/envconfig.go) it seems... this way we could integrate the file lookup directly into the package, so the `script.go` doesn't get even larger and more bloated (although not by much anyway). Also, we could fix the `SSH_IDENTITY_FILE` issue (_FILE_FILE looks quite bad to me and feels like a workaround) by just introducing a simple handle like `file` (similar  to `required` or `default`) if we want an env var to look for files and read then if present.

**Comment by m90** [comment]
> Also, we could fix the SSH_IDENTITY_FILE issue 

How'd you fix it other than not supporting passing this value in a secret/file? If this tool was to support the `_FILE` convention, I think _not_ applying it to everything would be pretty confusing and even worse than `_FILE_FILE`.

**Comment by MaxJa4** [comment]
Oh right, it makes more sense the other way around: disable it actively for variables which shall not be read as file, like with `SSH_IDENTITY_FILE`. Otherwise, always active.

**Comment by m90** [comment]
> disable it actively

I wouldn't even worry about disabling it. Considering it's not a secret value but a file location, very few people will be tempted of thinking it's even possible to pass the file location in `SSH_IDENTITY_FILE_FILE`. And if someone does, it still works as documented.

**Comment by MaxJa4** [comment]
Maybe something like `_FILE_PATH` is less confusing than `_FILE_FILE`? That would indicate clearer, that the intend here is to specify the path to a file, not necessarily its content as env variable.

**Comment by MaxJa4** [comment]
> True, but unfortunately _FILE is a convention used in many Docker images

All of these support the _FILE convention only for specific variables, not all.

> If you really want to do something about _FILE_FILE

I actually wouldn't want to change `SSH_IDENTITY_FILE` at all from the user's perspective to not break existing setups. Hence the suggestion to disable the auto-read functionality for this variable (e.g. with a flag like `no-read` or `filepath` or `value-only`), as reading a file here makes no sense and wouldn't confuse the user, as they will just provide the file path like normal anyway.

So something like this:

```go
// SSH_USER as value or SSH_USER_FILE as file
 SSHUser          string   `split_words:"true"`
// SSH_PASSWORD as value or SSH_PASSWORD_FILE as file
 SSHPassword      string   `split_words:"true"`
// SSH_IDENTITY_FILE as a value only:
 SSHIdentityFile  string   `split_words:"true" default:"/root/.ssh/id_rsa" filepath:"true"`
```

**Comment by m90** [comment]
I'm not entirely sure what we'd win from not enabling `_FILE` for all options, other than not having to cringe on seeing `_FILE_FILE`? Is there anything where it could become a non-aesthetic problem?

If we selectively do it, no matter if it's by disabling or enabling, we'll have to meticulously document it, and we'll have users showing up "Hey, can you also support `_FILE` for $MY_SETTING`, thanks.". Every new config option will have to decide whether it's something that should do `_FILE` or shouldn't. That's a lot of background noise.

If we enable it for everything, we have very little code to add/change right now and can document the feature by adding a single paragraph. Also, all future additions will support it out of the box.

**Comment by MaxJa4** [comment]
I just re-read your implementation example above and I think I might have misunderstood you with the `_FILE_FILE`.
I was interpreting that the user would need to supply `SSH_IDENTITY_FILE_FILE="/some/path/id_rsa"` now. But after reading the snippet above, it's just that the function would try to read `SSH_IDENTITY_FILE + _FILE` (hence the _FILE_FILE at this spot only) in the env vars and will most likely fail, then just pass it along as a variable, causing no issues. Right?

**Comment by m90** [comment]
Yes, exactly :)

**Comment by MaxJa4** [comment]
Alriiight, then forget everything I've said about the `_FILE_FILE` 😅
Then let's do the `_FILE` functionality for all variables with no disable/enable flag. Basically taking your snippet above and implementing it into `envconfig`.

So there's only one choice left:
- cleaning up the `envconfig` fork and using it like that
- keeping the fork as is, just making a PR basically to include the `.Lookup` function and using our fork in the meanwhile (or wait for a response in the main repo)
- implement the `_FILE` functionality directly into `envconfig` and move it (just envconfig.go as it seems) into the DVB repo.

**Comment by m90** [comment]
I can see pros and cons for all approaches and don't have a real preference. Do you have any?

If you also don't have one, let's do 2 as it still allows for moving to one of the other approaches later on. If you do have a preference, feel free to pick it, I approve of all of them.

**Comment by MaxJa4** [comment]
I have a slight preference for option 2 as well, since it's faster to implement and may be used by others if the PR gets merged.
If we need to do more fancy things with `envconfig` we can always go route 1 or 3 later, yes. But I don't see a benefit for the user right now to go the extra mile with 1 or 3.

**Comment by m90** [comment]
This is now included in [v2.32.0](https://github.com/offen/docker-volume-backup/releases/tag/v2.32.0)

**Comment by MaxJa4** [comment]
Great! Saw that docker hub is currently a little unstable, good thing that CI can be re-run ♻️
